server/main.lua

```diff
@@ -1,158 +1,434 @@
-    GiveKeys(src, plate)
-end)
+-- Returns plate from vehicle net id. Used for optimal code reuse
+---@param vehicleNetId number NetworkId of vehicle to get plate for
+---@return number? vehicle Vehicle entity id
+---@return string? plate Plate number of vehicle
+local function getVehiclePlateFromNetId(vehicleNetId)
+    local vehicle = NetworkGetEntityFromNetworkId(vehicleNetId) or nil
+    local plate = GetPlate(vehicle) or nil
+    return vehicle, plate
+end
 
-RegisterNetEvent('qb-vehiclekeys:server:breakLockpick', function(itemName)
-    local player = exports.qbx_core:GetPlayer(source)
-    if not player then return end
-    if not (itemName == "lockpick" or itemName == "advancedlockpick") then return end
-    exports.ox_inventory:RemoveItem(source, itemName, 1)
-end)
+-- Randomly breaks lockpicks from source. Toggles advanced or regular lockpicks depending on boolean
+---@param source number Player to remove lockpicks from
+---@param vehicleClass number Vehicle class number
+---@param isAdvanced boolean If true then breaks advanced lockpicks
+local function breakLockpick(source, vehicleClass, isAdvanced)
+    local chance = isAdvanced and config.removeAdvancedLockpickChance[vehicleClass] or config.removeNormalLockpickChance[vehicleClass]
+    local itemName = isAdvanced and 'advancedlockpick' or 'lockpick'
 
-RegisterNetEvent('qb-vehiclekeys:server:setVehLockState', function(vehNetId, state)
-    SetVehicleDoorsLocked(NetworkGetEntityFromNetworkId(vehNetId), state)
-end)
+    if math.random() <= chance then
+        exports.ox_inventory:RemoveItem(source, itemName, 1)
+    end
+end
 
-lib.callback.register('qbx-vehiclekeys:server:getVehicleKeys', function(source)
-    local citizenid = exports.qbx_core:GetPlayer(source).PlayerData.citizenid
-    local keysList = {}
-    for plate, citizenids in pairs (vehicleList) do
-        if citizenids[citizenid] then
-            keysList[plate] = true
+-- Remove Keys for all players
+---@param plate string Plate to be removed from vehicle key cache of all players
+local function removeKeysForAll(plate)
+    for _, playerId in ipairs(GetPlayers()) do
+        lib.callback('qbx_vehiclekeys:client:removeKeys', playerId, nil, plate)
+    end
+    vehicleList[plate] = nil
+end
+
+-- Remove keys from target id. If no id specified then removes all keys for a given plate.
+-- Export provided to support removing keys when garaging vehicles etc.
+---@param plate string Plate to be remove from vehicle key cache of target or all players
+---@param id? number Optional id to remove keys from
+local function removeKeys(plate, id)
+    local citizenid = nil
+    if id then
+        citizenid = exports.qbx_core:GetPlayer(id).PlayerData.citizenid
+    end
+
+    if vehicleList[plate] and vehicleList[plate][citizenid] and citizenid then
+        vehicleList[plate][citizenid] = nil
+        lib.callback('qbx_vehiclekeys:client:removeKeys', id, nil, plate)
+        for _, _ in pairs(vehicleList[plate]) do
```

Comment:
This is really for a corner case of if you manually spawn a vehicle and remove the keys from it. Suggestion on how to better handle this would be appreciated.